test/common/http/http2/codec_impl_test.cc

```diff
@@ -3223,6 +3264,54 @@ TEST_P(Http2CodecImplTest, ConnectTest) {
   driveToCompletion();
 }
 
+TEST_P(Http2CodecImplTest, ShouldWaitForDeferredBodyToProcessBeforeProcessingTrailers) {
+  // We must initialize before dtor, otherwise we'll touch uninitialized
+  // members in dtor.
+  initialize();
+
+  // Test only makes sense if we have defer processing enabled.
+  if (!defer_processing_backedup_streams_) {
+    return;
+  }
+
+  TestRequestHeaderMapImpl request_headers;
+  HttpTestUtility::addDefaultHeaders(request_headers);
+  EXPECT_CALL(request_decoder_, decodeHeaders_(_, false));
+  EXPECT_TRUE(request_encoder_->encodeHeaders(request_headers, false).ok());
+  driveToCompletion();
+
+  // Force the stream to buffer data at the receiving codec.
+  server_->getStream(1)->readDisable(true);
+  Buffer::OwnedImpl body(std::string(1024 * 1024, 'a'));
+  request_encoder_->encodeData(body, false);
+  driveToCompletion();
+
+  // Now re-enable the stream, and try dispatching trailers to the server.
+  // It should buffer those trailers until the buffered data is processed
+  // from the callback below.
+  auto* process_buffered_data_callback =
+      new NiceMock<Event::MockSchedulableCallback>(&server_connection_.dispatcher_);
+  EXPECT_FALSE(process_buffered_data_callback->enabled_);
+
+  server_->getStream(1)->readDisable(false);
+
+  EXPECT_TRUE(process_buffered_data_callback->enabled_);
+
+  // Trailers should be buffered by the codec since there is unprocessed body.
+  // Hence we shouldn't invoke decodeTrailers yet.
+  EXPECT_CALL(request_decoder_, decodeTrailers_(_)).Times(0);
+  request_encoder_->encodeTrailers(TestRequestTrailerMapImpl{{"trailing", "header"}});
+  driveToCompletion();
+
+  // Now invoke the deferred processing callback.
+  {
+    InSequence seq;
+    EXPECT_CALL(request_decoder_, decodeData(_, false));
+    EXPECT_CALL(request_decoder_, decodeTrailers_(_));
+    process_buffered_data_callback->invokeCallback();
```

Comment:
I think you could do with a bit more unit testing of corner cases here.  basics like back up of body with no fin doesn't send fin on resume.  back up of body with fin and no trailers, etc.

Reply:
added additional unit test